content/articles/flow-and-typescript-1.md

```diff
@@ -0,0 +1,190 @@
+  5: import { Route, IndexRoute } from "react-router";
+                                       ^^^^^^^^^^^^^^ exports of "react-router"
+```
+
+In the `routes.js` file we are importing `Route` from `react-router` but our interface file is completely empty.  We have two possibilities here, either we define `Route` to be `any` (if we are prototyping) or we annotate it properly.
+
+```js
+declare module "react-router" {
+  declare var Route: any;
+}
+// or
+type RouteProps = {
+  component: ReactClass<any, any, any>
+};
+
+type LinkProps = {
+  to: string
+};
+
+declare module "react-router" {
+  declare var Route: ReactClass<void, RouteProps, void>;
+  declare var IndexRoute: ReactClass<void, RouteProps, void>;
+  declare var Link: ReactClass<void, LinkProps, void>;
+}
+```
+
+Do remember that we just started using Flow so those declarations might be incorrect.
+
+Adding flow to the rest of the files is easy as well, with the occasional 3rd party library interface to write.
+
+Flow has a number of useful features but let's have a look at two in more details:
+
+### Import checking
+
+Importing an unknown property will result in a error, as shown in the following error:
+
+```
+src/app/components/DashboardRoute.js:9
+  9: import { getAllLists2 } from "../reducers/lists";
+              ^^^^^^^^^^^^ "getAllLists2" export of "../reducers/lists". Property not found in
+  9: import { getAllLists2 } from "../reducers/lists";
+                                  ^^^^^^^^^^^^^^^^^^^ exports of "../reducers/lists"
+```
+
+This is preferable over the default Javascript behaviour of assigning `undefined` to variables that aren't found in the imported library.
+
+### Union types
+
+Flow doesn't currently have enums (sum types) but allows some approximation with [disjoint union types](http://flowtype.org/blog/2015/07/03/Disjoint-Unions.html):
+
+```js
+type TodoAction = Add | Remove
+type Add = { type: 'Add', entry: string }
+type Remove = { type: 'Remove', id: number }
+```
+
+With this setup flow will complain if 1) the type doesn't match any of the defined types or 2) the structure doesn't match. E.g.
+
+```js
+let add: TodoAction = {type: 'Ad', entry: 'get milk'}
+```
+
+fails with:
+
+```
+ 17: let add: TodoAction = {type: 'Ad', entry: 'get milk'}
+                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ object literal. This type is incompatible with
+ 17: let add: TodoAction = {type: 'Ad', entry: 'get milk'}
+              ^^^^^^^^^^ union type
+```
+
+and
+
+```js
+let add: TodoAction = {type: 'Add'}
+```
+
+will fail with a similar error because the `entry` field is missing:
+
+```
+ 20: let add: TodoAction = {type: 'Add'}
+                           ^^^^^^^^^^^^^ object literal. This type is incompatible with
+ 20: let add: TodoAction = {type: 'Add'}
+              ^^^^^^^^^^ union type
+```
+
+Unfortunately this is no true sum type. E.g. the following example will work just fine even though `"omgbug"` is not valid for the `type` field:
```

Comment:
"Unfortunately this is no true sum type." <- I think this is correct but a bit unusual "This is no moon!", is there any word missing?